### DISCUSSION TITLE
`RoutingSession` equivalent for Core usage

### DISCUSSION THREAD

**Discussion by alphavector** [sqlalchemy/sqlalchemy#10097] [Usage Questions] [answered]
My postgres cluster consists of a master and several read-only replicas, and I want the right engine to be selected depending on the query type

In the documentation I saw an interesting implementation of using query routing to different replicas depending on the query type

```
class RoutingSession(Session):
                def get_bind(self, mapper=None, clause=None, **kw):
                    # within get_bind(), return sync engines
                    if mapper and issubclass(mapper.class_, MyOtherClass):
                        return engines['other'].sync_engine
                    elif self._flushing or isinstance(clause, (Update, Delete)):
                        return engines['leader'].sync_engine
                    else:
                        return engines[
                            random.choice(['follower1','follower2'])
                        ].sync_engine
```

But this only works for sessions

I'm trying to implement the same approach for Core usage, and so far I've gotten the simplest version:

```
import asyncio
import random

import sqlalchemy as sa
from sqlalchemy.ext.asyncio import AsyncEngine, create_async_engine

engines = {
    'leader': create_async_engine('sqlite+aiosqlite:///leader.db'),
    'other': create_async_engine('sqlite+aiosqlite:///other.db'),
    'follower1': create_async_engine('sqlite+aiosqlite:///follower1.db'),
    'follower2': create_async_engine('sqlite+aiosqlite:///follower2.db'),
}


class RoutingEngine:
    def __init__(self, engines: dict[str, sa.Engine | AsyncEngine]):
        self._engines = engines

    def __getattr__(self, name):
        engine = random.choice(tuple(self._engines.values()))
        return getattr(engine, name)

    async def dispose(self, close: bool = True) -> None:
        for engine in self._engines.values():
            await engine.dispose(close)


async def async_main() -> None:
    engine = RoutingEngine(engines)

    async with engine.begin() as conn:
        cur = await conn.execute(sa.select(10))
        print(cur.scalar())

    await engine.dispose()


asyncio.run(async_main())
```

This just selects a random engine from the submitted engines

Do I understand correctly that:
1) Choosing the correct engine by query type is not possible because we are calling the query from a `Connection` object, which is not associated with an `Engine` object?
2) It will not work when we open transactions, because when `engine.begin()` is called, BEGIN is immediately called on the database and we don't know in advance what query will be executed.
But for session it works, because all operations are first added to the internal map, and then sent to the database at once, and we can analyze by the type of query to which replica to send this or that request.

Right?

Are there any workarounds for implementing `RoutingEngine`?

**Comment by zzzeek** [comment, marked as answer]
you can do all those things by further implementing what the `Session` already does, that is, maintain a series of open connections / transactions at once.   the `Session` maintains this within a construct called `SessionTransaction` that has a dictionary of engine -> connection with transaction, which it pulls from for each query to choose the right currently active connection.

So when you say `session.begin()`, you get a new `SessionTransaction` that hasn't actually connected to anything yet.   then when you say `session.connection()`, that goes into the current `SessionTransaction`, selects the desired `Engine`, then pulls up a new `Connection` if one is not already present.

There's no reason you can't just use the `Session` here as a connection / transaction manager, and otherwise not use any other ORM functionality.
